### PR TITLE
Import files: Add multiple nested folders validation

### PR DESCRIPTION
**Pull request by pandah3**
## Description

Building on top of the validation from #575.

This adds validation for multiple subdirectories in the same level in a nested directory tree structure such as:

- Given directory: `uploads`
 - Nested directories: `2020`, `2019`, `2018`, `2017`
  - Nested directories: `01`, `02`, `03`, `04`, `05`, `06`
   - Individual files: `image.jpg`, `image2.jpg, etc.`

## Steps to Test

1. Check out PR.
1. Download this sample `uploads` directory and unzip:
[uploads.zip](https://github.com/Automattic/vip/files/5030577/uploads.zip)

1. Run `npm install`
1. Run `npm run build`
1. Run `node ./dist/bin/vip-import-validate-files.js uploads`
1. Verify messages:

- Folder structure validation
- File extension validation
- Filename character validation
- Intermediate images validation

----------------------------------------

log1
log 2
log 3
log 4

### CODE REVIEW

**Comment by pandah3** [comment]
I'd love some input on the output/logs that get displayed @SaxonF 🙇‍♀️ . Namely:

- The larger the client logs get, the messier it gets (the example output images in the description box are outputs from _partial_ import files)
- We need to coordinate the output from the media files validation (this PR & #575) and SQL file validation logs

Media file validation output:
validation simple

SQL file validation output:
Screen Shot 2020-08-05 at 3 03 20 PM

(Sorry for the blurry screenshot ^. Maybe @dchymko has a clearer one.)

**Comment by SaxonF** [comment]
Looks pretty awesome to me @pandah3! Nice one!

I like the grouping by folder as its similar to running a test suite. If we wanted to compress it a bit I guess we could group the "directory exists" success messages unless one of them fails? I also wonder if it would be helpful to include a summary at the end letting people know how close they are to passing?

image

These are just small things that could be added later if they make sense and shouldn't block a release.

File: src/lib/vip-import-validate-files.js [hunk 1]
```diff
@@ -187,80 +199,84 @@ export const findNestedDirectories = async directory => {
...
-	} else {
-		console.log( chalk.yellow( '✕' ), 'Recommended: Structure your WordPress media files into', chalk.magenta( '`uploads/YYYY`' ), 'directories' );
-		errors++;
-	}
+		/**
+			* Logging
+			*/
+
+		// Uploads folder
+		if ( uploadsIndex === 0 ) {
```
Comment (markowsiak) [inline src/lib/vip-import-validate-files.js:256]: I think the index parts are what didn't pass for me if I supply a path argument that isn't relative to the test execution (ie. `~/Downloads/upload`):

image

We might have to re-evaluate how we reference this.  Maybe we can skip validating what the index is of the uploads directory and just ensure that the year and month are at the upload directory index +1 and +2 respectively?  That's one way, but perhaps there's another if we have to validate the upload directory position?

File: src/lib/vip-import-validate-files.js [hunk 2]
```diff
@@ -142,40 +142,52 @@ const recommendAcceptableFileNames = () => {
...
+		nestedDirectories = nestedDirectories.filter( file => ! ( /(^|\/)\.[^\/\.]/g ).test( file ) );
+
+		nestedDirectories.forEach( dir => {
+			// Concatenate the file path of the parent directory with the nested directory
+			const filePath = path.join( directory, dir );
+			const statSync = fs.statSync( filePath ); // Get stats on the file/folder
+
+			// Keep looking for nested directories until we hit individual files
+			if ( statSync.isDirectory() ) {
+				findNestedDirectories( filePath );
```
Comment (markowsiak) [inline src/lib/vip-import-validate-files.js:175]: Should this have a maximum call depth to pass?
Or can customers have an unlimited number of nested folders inside of the `yyyy/mm/` directory?

**Review by markowsiak — COMMENTED** [review]
Looks good @pandah3 ⭐ 
I ran into an issue during test execution, but I think I found out why in the comments below.  Let me know if my suggestion is reasonable, or if the use-case of the path I supplied isn't required 👍

File: src/lib/vip-import-validate-files.js (hunk 2 above)
Comment (dchymko) [reply src/lib/vip-import-validate-files.js:175]: Somewhat of an edge case but not a bad idea. Maybe we should limit it to "X" and have a warning show if it's reached (and possibly a flag to override that number)

**Review by dchymko — DISMISSED** [review]
LGTM. That recursion thing is a minor edge case IMO from both the likelihood and also because this will run on a user's machine and not on affect sites, etc

File: src/lib/vip-import-validate-files.js [hunk 3]
```diff
@@ -411,3 +418,31 @@ export const logErrorsForIntermediateImages = obj => {
 	}
 	console.log( '------------------------------------------------------------' );
 };
+
+export const summaryLogs = ( folderErrorsLength, intImagesErrorsLength, fileTypeErrorsLength, filenameErrorsLength, totalFiles, totalFolders ) => {
```
Comment (elazzabi) [inline src/lib/vip-import-validate-files.js]: Minor feedback

In general, for functions with more than 2 arguments, it's hard to remember the order of arguments. Another way of doing this is by accepting an object as an argument with named attributes:

```js
export const summaryLogs = ( { folderErrorsLength, intImagesErrorsLength, fileTypeErrorsLength, filenameErrorsLength, totalFiles, totalFolders } )
```

When calling the function, it's easier than remembering the order:
```js
summaryLogs( {
  folderErrorsLength: folderValidation.length,
  intImagesErrorsLength: intermediateImages.tally,
  fileTypeErrorsLength: errorFileTypes.length,
  filenameErrorsLength: errorFileNames.length,
  totalFiles: files.length,
  totalFolders: nestedDirectories.length
} )
```

**Review by elazzabi — COMMENTED** [review]
Looks good. Left a minor feedback

File: src/lib/vip-import-validate-files.js (hunk 1 above)
Comment (pandah3) [reply src/lib/vip-import-validate-files.js:256]: Thank you for the feedback 🙏 ! I think in this instance, it's working as expected- if the supplied media folder is `Downloads/uploads`, the recommended structure is simply an `uploads` folder (without the `Downloads` folder). Maybe the wording needs to be clearer?

File: src/lib/vip-import-validate-files.js (hunk 2 above)
Comment (pandah3) [reply src/lib/vip-import-validate-files.js:175]: I'll do some more research into what the limits are and see if we should add one in a following PR!

File: src/lib/vip-import-validate-files.js (hunk 3 above)
Comment (pandah3) [reply src/lib/vip-import-validate-files.js]: Gah I meant to do this but forgot! Thank you! Fixed in 1abd712!

**Comment by pandah3** [comment]
Thank you everyone for the feedback! I added a summary per Saxon's suggestion! Will add any follow ups in a new PR.

summary

**Comment by pandah3** [comment]
Merging!

**Pull request merged**